File: engine-src/host/com/qwara/gomoku/engine/RapfiNative.java
package com.qwara.gomoku.engine;

/**
 * Desktop-JVM copy of the Android bridge object, used to smoke-test librapfi.so on the host.
 *
 * <p>The app module supplies the real class with the same package and class name; only the four
 * native methods below matter, and they must stay in sync with the JNI bridge in
 * {@code Rapfi/jni/rapfi_jni.cpp}.
 */
public final class RapfiNative {
    static {
        System.loadLibrary("rapfi");
    }

    private RapfiNative() {}

    /** Start (or restart) the protocol loop, resolving config and weights from {@code workDir}. */
    public static native void start(String workDir);

    /** Queue one command line for the engine. */
    public static native void write(String line);

    /** Block until one engine output line is available; null once the session ended and drained. */
    public static native String readLine();

    /** Whether the protocol loop thread is still alive. */
    public static native boolean isRunning();
}

File: app/src/main/java/com/qwara/go/engine/RapfiNative.kt
// SPDX-FileCopyrightText: 2026 Qwara-chan
// SPDX-License-Identifier: GPL-3.0-or-later

package com.qwara.go.engine

/**
 * Rapfi 引擎 JNI 桥。librapfi.so 在 App 进程内以独立线程运行 Piskvork 协议循环，
 * 输入/输出经 streambuf 重定向对接 Java 侧：
 *  - [start] 启动协议线程（幂等；END 退出后可再次调用以重建会话）；
 *  - [write] 写入一条协议命令；
 *  - [readLine] 阻塞读取一行引擎输出，会话结束且队列耗尽后返回 null；
 *  - [isRunning] 协议线程是否存活。
 */
object RapfiNative {
    init {
        System.loadLibrary("rapfi")
    }

    external fun start(workDir: String)
    external fun write(line: String)
    external fun readLine(): String?
    external fun isRunning(): Boolean
}
